core/src/main/java/org/apache/accumulo/core/zookeeper/ZooCache.java

```diff
@@ -143,65 +141,106 @@ public void process(WatchedEvent event) {
-        log.trace("{} removing watches for {} because {} accesses {}", cacheId, path, reason,
-            zcNode == null ? -1 : zcNode.getAccessCount());
-        zk.removeWatches(path, ZooCache.this.watcher, Watcher.WatcherType.Any, false);
-      } catch (InterruptedException | KeeperException | RuntimeException e) {
-        log.warn("{} failed to remove watches on path {} in zookeeper", cacheId, path, e);
+  public void addZooCacheWatcher(ZooCacheWatcher watcher) {
+    externalWatchers.add(requireNonNull(watcher));
+  }
+
+  // visible for tests
+  long getZKClientObjectVersion() {
+    return zk.getConnectionCounter();
+  }
+
+  private boolean handleZKConnectionChange() {
+    final long currentCount = getZKClientObjectVersion();
+    final long oldCount = zkClientTracker.get();
+    if (oldCount != currentCount) {
+      if (zkClientTracker.compareAndSet(oldCount, currentCount)) {
+        setupWatchers(watchedPaths);
       }
-    };
-    // Must register the removal listener using evictionListener inorder for removal to be mutually
-    // exclusive with any other operations on the same path. This is important for watcher
-    // consistency, concurrently adding and removing watches for the same path would leave zoocache
-    // in a really bad state. The cache builder has another way to register a removal listener that
-    // is not mutually exclusive.
-    Cache<String,ZcNode> cache =
-        Caches.getInstance().createNewBuilder(Caches.CacheName.ZOO_CACHE, false)
-            .expireAfterAccess(timeout).evictionListener(removalListerner).build();
-    nodeCache = cache.asMap();
-    log.trace("{} created new cache", cacheId, new Exception());
+      return true;
+    }
+    return false;
   }
 
-  private abstract static class ZooRunnable<T> {
+  // Called on construction and when ZooKeeper connection changes
+  synchronized void setupWatchers(Set<String> pathsToWatch) {
+
+    for (String left : pathsToWatch) {
+      for (String right : pathsToWatch) {
+        if (!left.equals(right) && left.contains(right)) {
+          throw new IllegalArgumentException(
+              "Overlapping paths found in paths to watch. left: " + left + ", right: " + right);
+        }
+      }
+    }
+
+    try {
+      zk.addPersistentRecursiveWatchers(pathsToWatch, watcher);
+      clear();
+      watchedPaths = Collections.unmodifiableNavigableSet(new TreeSet<>(pathsToWatch));
+    } catch (KeeperException | InterruptedException e) {
+      throw new RuntimeException("Error setting up persistent recursive watcher", e);
+    }
+
+  }
+
+  private boolean isWatchedPath(String path) {
+    // Check that the path is equal to, or a descendant of, a watched path
+    var floor = watchedPaths.floor(path);
+    return floor != null
+        && (floor.equals("/") || floor.equals(path) || path.startsWith(floor + "/"));
+  }
+
+  // Use this instead of Preconditions.checkState(isWatchedPath, String)
+  // so that we are not creating String unnecessarily.
```

Comment:
This comment feels like an apology to Guava for not using one of its methods, but Preconditions isn't the presumed default way to write this simple if statement, so we owe it no explanation.

That said, there is a version of Preconditions.checkState that takes a format string and format parameters, to construct the string only if the exception needs to be thrown. So, this comment isn't really necessary. We can either use that version of Preconditions.checkState (or maybe a newer version that takes a `Supplier<String>` if that exists), or just keep this method like it is, which generally speaks for itself without the comment.